source/Halibut/Util/CancelOnDisposeCancellationToken.cs

```diff
@@ -86,14 +89,20 @@ public void CancelAfter(TimeSpan timeSpan)
 
         /// <summary>
         /// Tasks supplied here will be awaited on in the dispose method after
-        /// the Token is cancelled and before the token is disposed. 
+        /// the Token is cancelled and before the token is disposed.
+        /// Thread safe: uses Interlocked.CompareExchange for initialization and ConcurrentBag for storage.
         /// </summary>
         /// <param name="tasksUsingToken"></param>
         public void AwaitTasksBeforeCTSDispose(params Task[] tasksUsingToken)
         {
+            if (tasks == null)
+            {
+                var newBag = new ConcurrentBag<Task>();
+                Interlocked.CompareExchange(ref tasks, newBag, null);
+            }
             foreach (var task in tasksUsingToken)
             {
-                tasks.Add(task);
+                tasks!.Add(task);
```

Comment:
nit: It would be great if we could avoid the `!` here, although I think an assertion in the `if` block might be the only realistic way.